### PR TITLE
Feature/pdpdevtool 2802 - Support for file upload in Node CLI

### PR DESCRIPTION
**Pull request by roctaviano**
Feature/pdpdevtool 2802 - Support for file upload in Node CLI

### CODE REVIEW

File: packages/node-cli/src/services/FileSystemService.js [hunk 1]
```diff
@@ -20,7 +20,20 @@ module.exports = class FileSystemService {
 
 		const availableDirectories = getDirectories(parentFolder);
 
-		return [parentFolder, ...availableDirectories];
+		return availableDirectories;
```
Comment (davidecorreu) [inline packages/node-cli/src/services/FileSystemService.js]: Actually now this function does exactly what it says, return the directories inside of the parentFolder, but it is used also in the ImportObjectsCommandGenerator.js , and there is expecting the parantFolder in the returned array. One possible fix can be adding there (ImportObjectsCommandGenerator.js) the '/Objects' folder as one of options to pick.

File: packages/node-cli/src/validation/FileCabinetValidator.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+const TEMPLATES_PATH = "/Templates";
+const TEMPLATES_EMAIL_TEMPLATES_PATH = "/Templates/E-mail Templates";
+const TEMPLATES_MARKETING_TEMPLATES_PATH = "/Templates/Marketing Templates";
+const WEB_SITE_HOSTING_FILES_PATH = "/Web Site Hosting Files";
+
+const UNRESTRICTED_PATHS = [SUITESCRIPTS_PATH, TEMPLATES_EMAIL_TEMPLATES_PATH, TEMPLATES_MARKETING_TEMPLATES_PATH, WEB_SITE_HOSTING_FILES_PATH];
+
+class FileCabinetValidator {
+	pathIsUnrestricted(path) {
+		//Templates folder is a special case, because it contains both restricted and unrestricted folders,
```
Comment (davidecorreu) [inline packages/node-cli/src/validation/FileCabinetValidator.js]: Can we upload a file to "/Templates" ?

**Comment by margabit** [comment]
Is this going to be published with the 20.1 release?

File: packages/node-cli/src/validation/FileCabinetValidator.js (hunk 2 above)
Comment (alii-gator) [reply packages/node-cli/src/validation/FileCabinetValidator.js]: Not directly under Templates..

**Comment by alii-gator** [comment]
Albert this was not part of Regression, so it will not be released in 1.0.0 version

File: packages/node-cli/src/commands/UploadFilesCommandGenerator.js [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+	SUCCESS: 'SUCCESS',
+	ERROR: 'ERROR',
+};
+
+const { validateArrayIsNotEmpty, showValidationResults } = require('../validation/InteractiveAnswersValidator');
+
+module.exports = class UploadFilesCommandGenerator extends BaseCommandGenerator {
+	constructor(options) {
+		super(options);
+		this._projectInfoService = new ProjectInfoService(this._projectFolder);
```
Comment (karl-anthony-ng) [inline packages/node-cli/src/commands/UploadFilesCommandGenerator.js]: Is projectInfoService used in here?

File: packages/node-cli/src/commands/UploadFilesCommandGenerator.js [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+		if (SDKOperationResultUtils.hasErrors(operationResult)) {
+			SDKOperationResultUtils.logResultMessage(operationResult);
+			SDKOperationResultUtils.logErrors(operationResult);
+			return;
+		}
+
+		if (Array.isArray(data)) {
+			const successfulUploads = data.filter(result => result.type === UPLOAD_FILE_RESULT_STATUS.SUCCESS);
+			const unsuccessfulUploads = data.filter(result => result.type === UPLOAD_FILE_RESULT_STATUS.ERROR);
+			if (Array.isArray(successfulUploads) && successfulUploads.length) {
```
Comment (karl-anthony-ng) [inline packages/node-cli/src/commands/UploadFilesCommandGenerator.js]: I think filter always returns an array, so no need to if it's an array here.

File: packages/node-cli/src/commands/UploadFilesCommandGenerator.js [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+				NodeUtils.println(TranslationService.getMessage(OUTPUT.FILES_UPLOADED), NodeUtils.COLORS.RESULT);
+				successfulUploads.forEach(result => {
+					NodeUtils.println(this._getFileCabinetRelativePath(result.file.path), NodeUtils.COLORS.RESULT);
+				});
+			}
+			if (Array.isArray(unsuccessfulUploads) && unsuccessfulUploads.length) {
```
Comment (karl-anthony-ng) [inline packages/node-cli/src/commands/UploadFilesCommandGenerator.js]: I think filter always returns an array, so no need to if it's an array here.

File: packages/node-cli/src/commands/UploadFilesCommandGenerator.js [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+	_formatOutput(operationResult) {
+		const { data } = operationResult;
+
+		if (SDKOperationResultUtils.hasErrors(operationResult)) {
+			SDKOperationResultUtils.logResultMessage(operationResult);
+			SDKOperationResultUtils.logErrors(operationResult);
+			return;
+		}
+
+		if (Array.isArray(data)) {
```
Comment (karl-anthony-ng) [inline packages/node-cli/src/commands/UploadFilesCommandGenerator.js]: There may not be a need to check if data is an array. I think from the operation result, it'll will always return data as an array. I'm not 100% sure though, please check.

File: packages/node-cli/src/commands/UploadFilesCommandGenerator.js [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+		const localFileCabinetFoldersChoices = localFileCabinetSubFolders.map(transformFoldersToChoicesFunc);
+
+		return [
+			{
+				message: TranslationService.getMessage(QUESTIONS.SELECT_FOLDER),
+				type: CommandUtils.INQUIRER_TYPES.LIST,
+				name: COMMAND_ANSWERS.SELECTED_FOLDER,
+				choices: localFileCabinetFoldersChoices,
+			},
```
[the remark this answers is not in the document]
Comment (roctaviano) [reply packages/node-cli/src/commands/UploadFilesCommandGenerator.js]: Agree.

File: packages/node-cli/src/services/FileSystemService.js (hunk 1 above)
Comment (roctaviano) [reply packages/node-cli/src/services/FileSystemService.js]: Agree. Adding the 'Objects' folder in ImportObjectsCommandGenerator.js

File: packages/node-cli/src/commands/UploadFilesCommandGenerator.js (hunk 3 above)
Comment (roctaviano) [reply packages/node-cli/src/commands/UploadFilesCommandGenerator.js]: No, this should be removed.

File: packages/node-cli/src/commands/UploadFilesCommandGenerator.js (hunk 4 above)
Comment (roctaviano) [reply packages/node-cli/src/commands/UploadFilesCommandGenerator.js]: Agree.

File: packages/node-cli/src/commands/UploadFilesCommandGenerator.js (hunk 6 above)
Comment (roctaviano) [reply packages/node-cli/src/commands/UploadFilesCommandGenerator.js]: This is to protect from the possibility that it could be null.

Comment (karl-anthony-ng) [reply packages/node-cli/src/commands/UploadFilesCommandGenerator.js]: But in this case, I think we know it'll always be an array right? But I'm not 100% sure so I'm ok if this check is necessary here.

File: packages/node-cli/src/commands/UploadFilesCommandGenerator.js [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+	ERROR: 'ERROR',
+};
+
+const { validateArrayIsNotEmpty, showValidationResults } = require('../validation/InteractiveAnswersValidator');
+
+module.exports = class UploadFilesCommandGenerator extends BaseCommandGenerator {
+	constructor(options) {
+		super(options);
+		this._fileSystemService = new FileSystemService();
+		this.localFileCabinetFolder = path.join(this._projectFolder, ApplicationConstants.FOLDERS.FILE_CABINET);
```
Comment (davidecorreu) [inline packages/node-cli/src/commands/UploadFilesCommandGenerator.js]: I think we can name it -> this._localFileCabinetFolder.

File: packages/node-cli/src/services/FileCabinetService.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+const TEMPLATES_EMAIL_TEMPLATES_PATH = '/Templates/E-mail Templates';
+const TEMPLATES_MARKETING_TEMPLATES_PATH = '/Templates/Marketing Templates';
+const WEB_SITE_HOSTING_FILES_PATH = '/Web Site Hosting Files';
+
+const UNRESTRICTED_PATHS = [SUITESCRIPTS_PATH, TEMPLATES_EMAIL_TEMPLATES_PATH, TEMPLATES_MARKETING_TEMPLATES_PATH, WEB_SITE_HOSTING_FILES_PATH];
+
+module.exports = class FileCabinetService {
+	constructor(fileCabinetAbsolutePath) {
+		this._fileSystemService = new FileSystemService();
+		this.fileCabinetAbsolutePath = fileCabinetAbsolutePath;
```
Comment (davidecorreu) [inline packages/node-cli/src/services/FileCabinetService.js]: this._fileCabinetAbsolutePath ?

File: packages/node-cli/src/services/FileCabinetService.js [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+				if (this._shouldEnterFolder(folder)) {
+					getFoldersRecursively(folder);
+				}
+			});
+		getFoldersRecursively(parentFolder);
+
+		return folders;
+	}
+
+	pathIsUnrestricted(path) {
```
Comment (davidecorreu) [inline packages/node-cli/src/services/FileCabinetService.js]: Maybe better isUnrestrictedPath(path) or _isUnrestrictedPath()?

**Pull request merged**